flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/component/DispatcherResourceManagerComponent.java

```diff
@@ -74,22 +79,46 @@
 			@Nonnull ResourceManager<?> resourceManager,
 			@Nonnull LeaderRetrievalService dispatcherLeaderRetrievalService,
 			@Nonnull LeaderRetrievalService resourceManagerRetrievalService,
-			@Nonnull WebMonitorEndpoint<?> webMonitorEndpoint) {
+			@Nonnull WebMonitorEndpoint<?> webMonitorEndpoint,
+			@Nonnull FatalErrorHandler fatalErrorHandler,
+			@Nonnull CompletableFuture<DispatcherGateway> dispatcherGatewayCompletableFuture) {
 		this.dispatcherRunner = dispatcherRunner;
 		this.resourceManager = resourceManager;
 		this.dispatcherLeaderRetrievalService = dispatcherLeaderRetrievalService;
 		this.resourceManagerRetrievalService = resourceManagerRetrievalService;
 		this.webMonitorEndpoint = webMonitorEndpoint;
+		this.fatalErrorHandler = fatalErrorHandler;
 		this.terminationFuture = new CompletableFuture<>();
 		this.shutDownFuture = new CompletableFuture<>();
 
 		registerShutDownFuture();
+		failOnPrematureTermination(dispatcherGatewayCompletableFuture);
 	}
 
 	private void registerShutDownFuture() {
 		FutureUtils.forward(dispatcherRunner.getShutDownFuture(), shutDownFuture);
 	}
 
+	private void failOnPrematureTermination(CompletableFuture<DispatcherGateway> dispatcherGatewayCompletableFuture) {
+		dispatcherGatewayCompletableFuture.whenComplete((dispatcher, throwable) -> {
+			if (dispatcher != null && dispatcher instanceof Dispatcher) {
+				CompletableFuture.anyOf(((Dispatcher) dispatcher).getTerminationFuture(), resourceManager.getTerminationFuture())
+					.whenComplete((ignored, t) -> {
```

Comment:
It may be a good idea to check whether t is indeed not null and log something accordingly if this is not the case. If it is null, then the process shut down without an error unexpectedly, which if I understand things correctly should not happen.

Reply:
We has modify `Dispatcher` and `ResourceManager` catch Exception to catch Throwable ，and then call `onFatalError` , so I think `DispatcherResourceManagerComponent#failOnPrematureTermination` is not very necessary, but @tillrohrmann  thinks it is safer

Reply:
I don't see how that relates to my comment. We will call the fatal error handler either way; the issue is that if t == null no reason for the shutdown will be logged-

Reply:
sorry, my understanding is wrong.